packages/libs/error-stack/src/ext/iter.rs

```diff
@@ -0,0 +1,338 @@
+    /// [`poll`]: Future::poll
+    fn change_context<C>(self, context: C) -> IteratorWithContext<Self, C>
+    where
+        C: Context;
+
+    /// Changes the [`Context`] of the [`Report`] inside the [`Result`] when calling
+    /// [`Iterator::next`]
+    ///
+    /// Applies [`Report::change_context`] to every [`Err`] variant in the iterator. For more
+    /// information, see the documentation for [`Report::change_context`].
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::change_context_lazy`]: crate::Report::change_context_lazy
+    /// [`poll`]: Future::poll
+    fn change_context_lazy<C, F>(self, context: F) -> IteratorWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: Fn() -> C;
+}
+
+impl<I: Iterator> IteratorExt for I
+where
+    I::Item: ResultExt,
+{
+    #[track_caller]
+    fn attach<A>(self, attachment: A) -> IteratorWithAttachment<Self, A> {
+        IteratorWithAttachment {
+            iterator: self,
+            context_or_attachment: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_lazy<A, F>(self, attachment: F) -> IteratorWithLazyAttachment<Self, F>
+    where
+        A: Send + Sync + 'static,
+        F: Fn() -> A,
+    {
+        IteratorWithLazyAttachment {
+            iterator: self,
+            context_or_attachment: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable<A>(self, attachment: A) -> IteratorWithPrintableAttachment<Self, A>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+    {
+        IteratorWithPrintableAttachment {
+            iterator: self,
+            context_or_attachment: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable_lazy<A, F>(
+        self,
+        attachment: F,
+    ) -> IteratorWithLazyPrintableAttachment<Self, F>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+        F: Fn() -> A,
+    {
+        IteratorWithLazyPrintableAttachment {
+            iterator: self,
+            context_or_attachment: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn change_context<C>(self, context: C) -> IteratorWithContext<Self, C>
+    where
+        C: Context,
+    {
+        IteratorWithContext {
+            iterator: self,
+            context_or_attachment: context,
+        }
+    }
+
+    #[track_caller]
+    fn change_context_lazy<C, F>(self, context: F) -> IteratorWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: Fn() -> C,
+    {
+        IteratorWithLazyContext {
+            iterator: self,
+            context_or_attachment: context,
+        }
+    }
+}
+
+macro_rules! impl_iterator_adaptor {
+    (
+        $name:ident,
+        $method:ident,
+        $bound:ident
+        $(+ $bounds:ident)*
+        $(+ $lifetime:lifetime)*,
+        $output:ty
+    ) => {
+        #[doc=concat!("The adaptor returned by [`IteratorExt::", stringify!($method), "']")]
+        pub struct $name<I, A> {
```

Comment:
Let's use `T` instead
```suggestion
        pub struct $name<I, T> {
```